metals/src/main/scala/scala/meta/metals/lsp/TextDocumentService.scala

```diff
@@ -0,0 +1,171 @@
+package scala.meta.metals.lsp
+
+import java.util
+import java.util.concurrent.CompletableFuture
+
+import scala.meta.internal.metals.HoverExtParams
+
+import org.eclipse.lsp4j._
+import org.eclipse.lsp4j.jsonrpc.messages.{Either => JEither}
+import org.eclipse.lsp4j.jsonrpc.services.JsonNotification
+import org.eclipse.lsp4j.jsonrpc.services.JsonRequest
+import org.eclipse.{lsp4j => l}
+
+/**
+ * Interface which describes text document LSP requests and notifications which are
+ * implemented by Metals.
+ *
+ * It's equivalent to [[org.eclipse.lsp4j.services.TextDocumentService]] with some return types changed.
+ *
+ * Each method has a default implementation which throws UnsupportedOperationException. Throwing in this context is ok because:
+ * - the default implementation should never be called, because the method is always overridden in [[scala.meta.internal.metals.MetalsLanguageServer]]
+ * - lsp4j wraps each method in try catch and returns an error response to the client
+ */
+trait TextDocumentService {
```

Comment:
So this not implements all possible methods from LSP spec? Even if it isn't yet implemented by Metals?

Reply:
No, it contains only those which are implemented by Metals